src/julia.jl

```diff
@@ -53,54 +53,63 @@ function pairwise(f::Function, a, ::Type{Symmetric})
     n = length(a)
     firstval = f(first(a), first(a))
     r = Symmetric(Matrix{typeof(firstval)}(n, n))
-    pairwise!(f, a, r)
+    pairwise_internal!(f, a, r, firstval)
     r
 end
 
 """
     pairwise!(f::Function, a, r)
 
 Apply the function `f` to all pairwise combinations of elements from `a` and
-storing the results in `r`. Setting keyword `symmetric = true` will only calculate
-each variable combination once for symmetric functions. This is useful if `f` is
-expensive.
+storing the results in `r`. If `r` is a `Symmetric` the function will only
+calculate each variable combination once, which is useful if `f` is expensive
+and `f(x,y) == f(y,x)`.
 """
 function pairwise!(f::Function, a, r::AbstractMatrix)
     n = length(a)
     size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r ($(size(r)), should be $((n, n)))"))
-    pairwise_internal!(f, a, r)
+    pairwise_internal!(f, a, r, f(first(a), first(a)))
 end
 
-function pairwise_internal!(f, a, r::AbstractMatrix)
+function pairwise_internal!{T}(f, a, r::AbstractMatrix{T}, firstval::T)
+    r[1,1] = firstval
     @inbounds for j in eachindex(a)
         aj = a[j]
         for i in eachindex(a)
-            r[i,j] = f(a[i], aj)
+            if i != 1 || j != 1
```

Comment:
The problem with trying to make it work on `Dict` is that the iteration order of a `Dict` is undefined, so it would be hard to actually use the resulting `r` array reliably.